File: LegendEditor/Assets/Scripts/Game/SceneMap/MapTile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;

public class MapTile : GameBase
{
	public int mIndex;
	public short mBngImgIdx;	// 背景图索引
	public short mMidImgIdx;	// 补充背景图索引
	public short mObjImgIdx;	// 对象图索引
	public byte mDoorIdx;		// 门索引
	public byte mDoorOffset;	// 门偏移
	public byte mAniFrame;		// 动画帧数
	public byte mAniTick;		// 动画跳帧数
	public byte mObjFileIdx;	// 资源文件索引
	public byte mLight;			// 亮度
	public bool mHasBng;		// 是否有背景图(在热血传奇2地图中，背景图大小为4个地图块，具体到绘制地图时则表现在只有横纵坐标都为双数时才绘制)
	public bool mCanWalk;		// 是否可行走(站立)
	public bool mHasMid;		// 是否有补充图
	public bool mHasObj;		// 是否有对象图
	public bool mCanFly;		// 是否可以飞越
	public bool mHasDoor;		// 是否有门
	public bool mDoorOpen;		// 门是否开启	 
	public bool mHasAni;		// 是否有动画
	public MapTile(int index)
	{
		mIndex = index;
	}
	public void parseTile(byte[] buffer, ref int offset)
	{
		Serializer serializer = new Serializer(buffer);
		serializer.setIndex(offset);
		serializer.read(ref mBngImgIdx);
		serializer.read(ref mMidImgIdx);
		serializer.read(ref mObjImgIdx);
		serializer.read(ref mDoorIdx);
		serializer.read(ref mDoorOffset);
		serializer.read(ref mAniFrame);
		serializer.read(ref mAniTick);
		serializer.read(ref mObjFileIdx);
		serializer.read(ref mLight);
		offset = serializer.getIndex();
		
		mHasBng = BinaryUtility.getHightestBit(mBngImgIdx) == 1;
		BinaryUtility.setHighestBit(ref mBngImgIdx, 0);
		mHasMid = BinaryUtility.getHightestBit(mMidImgIdx) == 1;
		BinaryUtility.setHighestBit(ref mMidImgIdx, 0);
		mHasObj = BinaryUtility.getHightestBit(mObjImgIdx) == 1;
		BinaryUtility.setHighestBit(ref mObjImgIdx, 0);
		mCanWalk = (!mHasBng && !mHasObj);
		mCanFly = !mHasObj;
		mDoorOpen = BinaryUtility.getHightestBit(mDoorOffset) == 1;
		mHasDoor = BinaryUtility.getHightestBit(mDoorIdx) == 1;
		mHasAni = BinaryUtility.getHightestBit(mAniFrame) == 1;
		BinaryUtility.setHighestBit(ref mAniFrame, 0);
	}
}

File: LegendDataEditor/LegendDataEditor/SceneMap/MapTile.cpp
#include "MapTile.h"
#include "txSerializer.h"
#include "MapTileSimple.h"

void MapTile::parseTile(char* buffer, int bufferSize, int& offset)
{
	txSerializer serializer(buffer, bufferSize);
	serializer.setIndex(offset);
	serializer.read(mBngImgIdx);
	serializer.read(mMidImgIdx);
	serializer.read(mObjImgIdx);
	serializer.read(mDoorIdx);
	serializer.read(mDoorOffset);
	serializer.read(mAniFrame);
	serializer.read(mAniTick);
	serializer.read(mObjFileIdx);
	serializer.read(mLight);
	offset = serializer.getIndex();

	mOriginBngImgIdx = mBngImgIdx;
	mOriginMidImgIdx = mMidImgIdx;
	mOriginObjImgIdx = mObjImgIdx;
	mOriginAniFrame = mAniFrame;
	bool hasBng = GET_HIGHEST_BIT(mBngImgIdx) == 1;
	bool hasObj = GET_HIGHEST_BIT(mObjImgIdx) == 1;
	mCanWalk = (!hasBng && !hasObj);
	mCanFly = !hasObj;
	mDoorOpen = GET_HIGHEST_BIT(mDoorOffset) == 1;
	mHasDoor = GET_HIGHEST_BIT(mDoorIdx) == 1;
	mHasAni = GET_HIGHEST_BIT(mAniFrame) == 1;
	SET_HIGHEST_BIT(mBngImgIdx, 0);
	SET_HIGHEST_BIT(mMidImgIdx, 0);
	SET_HIGHEST_BIT(mObjImgIdx, 0);
	SET_HIGHEST_BIT(mAniFrame, 0);
}

void MapTile::saveTile(txSerializer* serializer)
{
	serializer->write(mOriginBngImgIdx);
	serializer->write(mOriginMidImgIdx);
	serializer->write(mOriginObjImgIdx);
	serializer->write(mDoorIdx);
	serializer->write(mDoorOffset);
	serializer->write(mOriginAniFrame);
	serializer->write(mAniTick);
	serializer->write(mObjFileIdx);
	serializer->write(mLight);
}

void MapTile::toSimple(MapTileSimple* simpleTile)
{
	simpleTile->mBngImgIdx = mBngImgIdx;
	simpleTile->mMidImgIdx = mMidImgIdx;
	simpleTile->mObjImgIdx = mObjImgIdx;
	simpleTile->mObjFileIdx = mObjFileIdx;
	simpleTile->mCanWalk = mCanWalk;
}

string MapTile::toString()
{
	string str;
	str += StringUtility::intToString(mOriginBngImgIdx);
	str += " ";
	str += StringUtility::intToString(mOriginMidImgIdx);
	str += " ";
	str += StringUtility::intToString(mOriginObjImgIdx);
	str += " ";
	str += StringUtility::intToString(mDoorIdx);
	str += " ";
	str += StringUtility::intToString(mDoorOffset);
	str += " ";
	str += StringUtility::intToString(mOriginAniFrame);
	str += " ";
	str += StringUtility::intToString(mAniTick);
	str += " ";
	str += StringUtility::intToString(mObjFileIdx);
	str += " ";
	str += StringUtility::intToString(mLight);
	return str;
}

void MapTile::parseString(const string& str)
{
	myVector<string> vec;
	StringUtility::split(str.c_str(), " ", vec);
	if (vec.size() != 9)
	{
		cout << "错误:地砖的文本数据错误,下标:" << mIndex << endl;
		return;
	}
	mBngImgIdx = StringUtility::stringToInt(vec[0]);
	mMidImgIdx = StringUtility::stringToInt(vec[1]);
	mObjImgIdx = StringUtility::stringToInt(vec[2]);
	mDoorIdx = StringUtility::stringToInt(vec[3]);
	mDoorOffset = StringUtility::stringToInt(vec[4]);
	mAniFrame = StringUtility::stringToInt(vec[5]);
	mAniTick = StringUtility::stringToInt(vec[6]);
	mObjFileIdx = StringUtility::stringToInt(vec[7]);
	mLight = StringUtility::stringToInt(vec[8]);

	mOriginBngImgIdx = mBngImgIdx;
	mOriginMidImgIdx = mMidImgIdx;
	mOriginObjImgIdx = mObjImgIdx;
	mOriginAniFrame = mAniFrame;
	bool hasBng = GET_HIGHEST_BIT(mBngImgIdx) == 1;
	bool hasObj = GET_HIGHEST_BIT(mObjImgIdx) == 1;
	mCanWalk = (!hasBng && !hasObj);
	mCanFly = !hasObj;
	mDoorOpen = GET_HIGHEST_BIT(mDoorOffset) == 1;
	mHasDoor = GET_HIGHEST_BIT(mDoorIdx) == 1;
	mHasAni = GET_HIGHEST_BIT(mAniFrame) == 1;
	SET_HIGHEST_BIT(mBngImgIdx, 0);
	SET_HIGHEST_BIT(mMidImgIdx, 0);
	SET_HIGHEST_BIT(mObjImgIdx, 0);
	SET_HIGHEST_BIT(mAniFrame, 0);
}